### PR TITLE
MAINT: Update to spin 0.12

### PR DESCRIPTION
**Pull request by stefanv**

### CODE REVIEW

**Comment by stefanv** [comment]
Can someone check if https://github.com/numpy/numpy/pull/27371/commits/f6680e44a1bbd0085fc012c9acfefa95c6b58dc1 is acceptable?

**Comment by charris** [comment]
Doctests are failing badly.

File: numpy/f2py/tests/test_abstract_interface.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import pytest
-from . import util
+from numpy.f2py.tests import util
```
Comment (mattip) [inline numpy/f2py/tests/test_abstract_interface.py:2]: What is the reason for these changes?

Comment (stefanv) [reply numpy/f2py/tests/test_abstract_interface.py:2]: If you look at the CI failure in the commit before that one, it fails on one platform to import util from the fortran tests. I'm not sure why it only fails on the one platform.

Comment (mattip) [reply numpy/f2py/tests/test_abstract_interface.py:2]: It doesn't fail with the current spin, right? So what changed with spin? Can you debug this locally? It seems like an unnecessary change.

Comment (stefanv) [reply numpy/f2py/tests/test_abstract_interface.py:2]: It wasn't an unnecessary change, per se:

https://github.com/scientific-python/spin/issues/209

`pytest` test discovery is quite complicated, unfortunately, and we have multiple ways in which people organize source (`src` vs `numpy`) and specify which they wish to run. I'll take another look next week, but the fact that none of the other projects ran into issues is what made me attempt a workaround in NumPy.

Comment (mattip) [reply numpy/f2py/tests/test_abstract_interface.py:2]: I would prefer we have a local override for the `spin test` command that will work around the shortcomings of the global `spin test` pytest invocation.

**Comment by lagru** [comment]
Is it possible that [spin invoking pytest with `--import-mode=importlib`](https://github.com/numpy/numpy/actions/runs/10965124616/job/30450225884#step:11:25) is why the relative imports are suddenly failing? [Pytests documentation](https://docs.pytest.org/en/stable/explanation/pythonpath.html#import-modes) states that for that mode

> Test modules can’t import each other.

Without checking every file, it seems to me that this is the only place where `spin test` is used; other passing workflows seem to use `pytest` directly.

That [option is added by spin if not already given](https://github.com/scientific-python/spin/blob/6669420f626059113980508d214728f4b3fe85c7/spin/cmds/meson.py#L540-L541) and was added in https://github.com/scientific-python/spin/commit/e8e8943384fe973fe6b287a42954735d179f1edb (v11-v12).

**Comment by stefanv** [comment]
@lagru Can you replicate the problem on your system? I can't.

**Comment by rgommers** [comment]
I can reproduce the problem with `spin` 0.12 for the `numpy.f2py` tests, as well as for SciPy in the in-progress work to move to `spin` (xref https://github.com/scientific-python/spin/pull/238#discussion_r1834428942).

@lagru's explanation about `--import-mode` above seems very plausible. 

> I would prefer we have a local override for the `spin test` command that will work around the shortcomings of the global `spin test` pytest invocation.

I tend to agree. Forbidding relative imports is an annoying limitation, and the failure mode is very obscure. So far we've been able to support test layouts with relative imports just fine in both NumPy (with older `spin`) and with SciPy (with `dev.py`), while also supporting editable installs + direct `pytest` invocation.

It seems to me like it should be feasible to handle this within `spin` either automatically or at most by adding a single config (e.g. `path-to-package-root = 'numpy'`) in `pyproject.toml`.

**Comment by stefanv** [comment]
Will open a new PR for spin 0.13.
